Send / receive zero values for uncoupled pixel

niklaswr opened the pull request:
- force CLM to send zero values for uncoupled pixel, as missing values like 9999 could crash other components if they receive those.
- force ParFlow to receive zero values for uncoupled pixel, as missing values like 9999 could crash ParFlow, if they are treated as valid forcing.

jjokella commented:
I am not familiar with this coupling, so my input is limited.

I have this general question:  By hard-coding the zero in this coupling step, could some model configurations run that one would rather have crashing?

niklaswr commented:
Good question, as Im also fan of `let a program crash, if something went wrong` :)  
However, the default behavior is to send zeros. Just CLM did send `-999999._r8`, and even this seems to be a test or workaround only, as there is a code line to send zeros, which gets overwritten in the next line, which looks quiet strange to me.  
Check [the commit changes](https://github.com/HPSCTerrSys/TSMP/pull/160/commits/8f518867b5c09987128690c842882cf361b891ac#diff-99b426061812ba87b99654d3d8294023171dd0d7ac3d2054f2afc6fc4cb599baL166-L167) for reference.  

So to my best knowledge, the hard-coded sending zeros should be the default, and if some configurations does crash with this, they are build on a bug and should be fixed, not a reason to not change to send zeros.

s-poll commented:
> I am not familiar with this coupling, so my input is limited.
> 
> I have this general question: By hard-coding the zero in this coupling step, could some model configurations run that one would rather have crashing?

In theory it is possible that the model won't crash anymore, when the time between the model component and oasis do not match. However, this should not happen if users do not modify code and use the established work flows. So to the best of my knowledge it would be a decisions between treatment of fields for masked values and treatment of fields for time issues between oasis and model component (even though this might not happen in practice).

I am also not so familiar with the masked coupling between CLM3.5 and ParFlow. Is there a case (multiple cases) where the masked values are treated as valid input in ParFlow and why are the masked values not consistent?

niklaswr commented:
I dont know why masked values are not treated consistently, but as I read the src-code this inconsistency (sending `-999999._r8` instead of `0._r8`) is an artifact only, as already mentioned  in the [commen above](https://github.com/HPSCTerrSys/TSMP/pull/160#issuecomment-1445915928).

Further to my best understanding CLM3.5 and ParFlow does treat every received value as valid input. So in fact there is no masking in a common sense, why it is important to not send nonphysical values.

niklaswr commented:
Just updated some formatting. In particular, removed start and end comments indicating changes made. Also removed author name and date of change, as this is tracked by git.

Are there still any concerns about this PR?
If not, I would like to merge it in the next few days, as this change is required by the new EUR-11 domain, where 'active' lakes are used in ParFlow.

s-poll commented:
Thank you for cleaning up the PR src, much better to read now. I took a look into the PR. 

I think sending `-999999._r8` instead of sending `0._r8` was not just an artifact but was intended (in fact only oas3 had `0._r8` in the src, which is not used by default - I am even not sure if oasis3 (without mct) is compiled at JSC). 

If I understand your problem correctly (please correct if not), you want to use quantities of CLM at grid points, which are masked/inactive for the CLM component but active for the ParFlow component (e.g. lakes). To my opinion a much more robust solution would be to indicate/modify these grid points in the ParFlow rcv routine.

@DCaviedesV What is your opinion on the PR?

DCaviedesV commented:
Well, I'm not sure I fully grasp the implications. 
I'm not sure what "default" means in @niklaswr's  [comment above](https://github.com/HPSCTerrSys/TSMP/pull/160#issuecomment-1445915928), or what "artifact" means in following comments. 
However, the arguments for consistenty from @niklaswr make sense to me.

@s-poll what would be the intention of using `-999999._r8` instead of `0.r8? Is it strictly indicating that there is masking happening? Or alternatively, what would be the implications of setting all to `0.r8`? 

How would the masking be enforced on the ParFLow `rcv` routine? Setting values to `-9999999`? (assuming you are right in what @niklaswr wants to do, and that the inverse use CLM active /PF inactive is of no interest).

Sorry, more questions than answers.

s-poll commented:
@DCaviedesV The default of sending fields (e.g. `-999999._r8`) indicate that values, which are not updated by CLM, can be identified as those. The reason for not updating could be masking but also other reason where not update took place, e.g. falsely synchronized time steps between oasis and component models. The implication to set this to `0._r8` would be that any user might not recognize if the CLM were not updated the fields (e.g. by synchronized masks, timing, etc. ) as one can not distinguish between physical meaningful results and the `0._r8` send to ParFlow, which are in the physical meaningful range. This is not so of importance for established setups, but might be for new ones.

niklaswr commented:
@s-poll: correct. I want lakes to be active in ParFlow, but inactive / masked in CLM3.5.

I agree with you, that changing the ParFlow `rcv` routine is most intuitive, and I have to agree, that sending nonphysical values for inactive / masked pixel is straight and very clean as well. Better crash a simulation if something went wrong instead of silently running through. 

I implemented the change as it is, because ParFlow is masking incoming values based on its own mask (`.pfsol`). This is working fine as long as CLM and ParFlow does share the same mask, but as I want leaks to be active, those pixel are not masked in ParFlow anymore and the nonphysical values send from CLM reaching ParFlow. Because of this I would have to implement another mask in ParFlow `rcv` routine, checking for the hard-coded values send by CLM for inactive / masked pixel. 
This would be fine if the ParFlow `rcv` routine is controlled by TSMP via patching, but the ParFlow `rcv` routine is part of the official ParFlow repo since some time. Thus I would have to change the official ParFlow repo with a hard-coded value used in TSMP, what is kind of confusing, not straight, and maintain intense. 
That is why I implemented the change to CLM `snd` routine, which is controlled by TSMP.

So I do see the following two options now (after bilateral discussing with @s-poll):
i) We stick with the current solution, which I am no longer 100% convinced of myself.
ii) There is a `rcv` routine in official ParFlow repo called [receive_fld2_clm.F90](https://github.com/parflow/parflow/blob/master/pfsimulator/amps/oas3/receive_fld2_clm.F90), which seems to be used by TSMP only. If so I do not have any doubt in implementing hard-coded values there. So I can test changint the ParFlow `rcv` routine there and request for a PR to official ParFlow repo.
